File: tests/python_align/cuda/test_reduction.py
"""Reduction alignment tests — Insight CUDA vs NumPy.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python \\
    LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \\
    python -m pytest tests/python_align/cuda/test_reduction.py -v
"""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
    from numpy.testing import assert_allclose
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)

# Skip if CUDA not available
try:
    ins.init()
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)

GPU = ins.GPUPlace(0)
CPU = ins.CPUPlace()


def to_gpu(np_arr):
    return ins.from_numpy(np_arr).to(GPU)


def to_numpy(gpu_arr):
    return gpu_arr.to(CPU).numpy()


class TestReductionAlignmentGPU:
    """Insight reduction ops vs NumPy."""

    @pytest.fixture
    def data_1d(self):
        return np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=np.float64)

    @pytest.fixture
    def data_2d(self):
        return np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float64)

    def test_sum(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.sum(x).numpy(), np.sum(data_1d))

    def test_mean(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.mean(x).numpy(), np.mean(data_1d))

    def test_max(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.max(x).numpy(), np.max(data_1d))

    def test_min(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.min(x).numpy(), np.min(data_1d))

    def test_sum_axis(self, data_2d):
        x = ins.from_numpy(data_2d)
        assert_allclose(ins.sum(x, axis=0).numpy(), np.sum(data_2d, axis=0))
        assert_allclose(ins.sum(x, axis=1).numpy(), np.sum(data_2d, axis=1))

    def test_mean_axis(self, data_2d):
        x = ins.from_numpy(data_2d)
        result = ins.mean(x, axis=0)
        expected = np.mean(data_2d, axis=0)
        # Insight may keep reduced dimension; squeeze to match NumPy
        assert_allclose(result.numpy().squeeze(), expected)

    def test_argmax(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.argmax(x).numpy(), np.argmax(data_1d))

    def test_argmin(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.argmin(x).numpy(), np.argmin(data_1d))

    def test_prod(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.prod(x).numpy(), np.prod(data_1d), rtol=1e-5)

    def test_var(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.var(x).numpy(), np.var(data_1d), rtol=1e-5)

    def test_std(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.std(x).numpy(), np.std(data_1d), rtol=1e-5)

    def test_cumsum(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.cumsum(x, 0).numpy(), np.cumsum(data_1d))


class TestReductionExtendedGPU:
    """Extended reduction ops vs NumPy."""

    @pytest.fixture
    def data_1d(self):
        return np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=np.float64)

    @pytest.fixture
    def data_2d(self):
        return np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float64)

    def test_cumprod(self, data_1d):
        x = ins.from_numpy(data_1d)
        assert_allclose(ins.cumprod(x, 0).numpy(), np.cumprod(data_1d), rtol=1e-5)

    def test_cummax(self, data_1d):
        x = ins.from_numpy(data_1d)
        result = ins.cummax(x, 0)
        expected = np.maximum.accumulate(data_1d)
        assert_allclose(result.numpy(), expected)

    def test_cummin(self, data_1d):
        x = ins.from_numpy(data_1d)
        result = ins.cummin(x, 0)
        expected = np.minimum.accumulate(data_1d)
        assert_allclose(result.numpy(), expected)

    def test_median_odd(self):
        x_np = np.array([3, 1, 4, 1, 5], dtype=np.float64)
        x = ins.from_numpy(x_np)
        assert_allclose(ins.median(x).numpy(), np.median(x_np))

    def test_median_even(self):
        x_np = np.array([3, 1, 4, 1, 5, 9], dtype=np.float64)
        x = ins.from_numpy(x_np)
        assert_allclose(ins.median(x).numpy(), np.median(x_np))

    def test_quantile(self, data_1d):
        x = ins.from_numpy(data_1d)
        result = ins.quantile(x, 0.5)
        assert_allclose(result.numpy(), np.quantile(data_1d, 0.5), rtol=1e-5)

    def test_quantile_75(self, data_1d):
        x = ins.from_numpy(data_1d)
        result = ins.quantile(x, 0.75)
        assert_allclose(result.numpy(), np.quantile(data_1d, 0.75), rtol=1e-5)

    def test_count_nonzero(self):
        x_np = np.array([0, 1, 0, 3, 0, 5], dtype=np.float64)
        x = ins.from_numpy(x_np)
        result = ins.count_nonzero(x)
        # count_nonzero returns scalar
        assert_allclose(result.numpy(), np.count_nonzero(x_np))

    def test_count_nonzero_2d(self):
        x_np = np.array([[0, 1], [2, 0], [0, 3]], dtype=np.float64)
        x = ins.from_numpy(x_np)
        result = ins.count_nonzero(x)
        assert_allclose(result.numpy(), np.count_nonzero(x_np))

    def test_nansum(self):
        x_np = np.array([1.0, np.nan, 3.0, np.nan, 5.0], dtype=np.float64)
        x = ins.from_numpy(x_np)
        assert_allclose(ins.nansum(x).numpy(), np.nansum(x_np), rtol=1e-8)

    def test_nanmean(self):
        x_np = np.array([1.0, np.nan, 3.0, np.nan, 5.0], dtype=np.float64)
        x = ins.from_numpy(x_np)
        assert_allclose(ins.nanmean(x).numpy(), np.nanmean(x_np), rtol=1e-8)

    def test_nanstd(self):
        x_np = np.array([1.0, np.nan, 3.0, np.nan, 5.0], dtype=np.float64)
        x = ins.from_numpy(x_np)
        assert_allclose(ins.nanstd(x).numpy(), np.nanstd(x_np), rtol=1e-5)

    def test_nanvar(self):
        x_np = np.array([1.0, np.nan, 3.0, np.nan, 5.0], dtype=np.float64)
        x = ins.from_numpy(x_np)
        assert_allclose(ins.nanvar(x).numpy(), np.nanvar(x_np), rtol=1e-5)

    def test_any(self):
        x_np = np.array([False, False, True, False])
        x = ins.from_numpy(x_np)
        assert_allclose(ins.any(x).numpy(), np.any(x_np))

    def test_all(self):
        x_np = np.array([True, True, True, False])
        x = ins.from_numpy(x_np)
        assert_allclose(ins.all(x).numpy(), np.all(x_np))

    def test_sem(self, data_1d):
        x = ins.from_numpy(data_1d)
        # standard error of the mean = std / sqrt(n)
        expected = np.std(data_1d) / np.sqrt(len(data_1d))
        assert_allclose(ins.sem(x).numpy(), expected, rtol=1e-5)

File: tests/cuda/julia/test_reduction.jl
# Reduction CUDA tests — aligned with C++ test_reduction.cpp
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_reduction.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0; failed = 0
function check(name, cond)
    global passed, failed
    if cond; passed += 1; else; failed += 1; println("FAIL: $name"); end
end

println("=== Reduction CUDA ===")

# sum
a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0]), Insight.GPUPlace(0))
s = Insight.sum(a)
check("sum", Insight.numel(s) == 1)

# sum_axis
b = Insight.to_array(Insight.from_data(Float64[1 2 3; 4 5 6]), Insight.GPUPlace(0))
s = Insight.sum(b, axis=1)
check("sum_axis", Insight.numel(s) == 3)

# mean
m = Insight.mean(a)
check("mean", Insight.numel(m) == 1)

# mean_axis
m = Insight.mean(b, axis=1)
check("mean_axis", Insight.numel(m) == 2)

# max
c = Insight.to_array(Insight.from_data([3.0, 1.0, 4.0, 1.0, 5.0]), Insight.GPUPlace(0))
m = Insight.max(c)
check("max", Insight.numel(m) == 1)

# min
m = Insight.min(c)
check("min", Insight.numel(m) == 1)

# prod
p = Insight.prod(a)
check("prod", Insight.numel(p) == 1)

# argmax
d = Insight.to_array(Insight.from_data([1.0, 5.0, 3.0, 2.0]), Insight.GPUPlace(0))
m = Insight.argmax(d)
check("argmax", Insight.numel(m) == 1)

# argmin
m = Insight.argmin(d)
check("argmin", Insight.numel(m) == 1)

# cummax
m2 = Insight.to_array(Insight.from_data(Float32[1 2 3; 4 5 6]), Insight.GPUPlace(0))
check("cummax", Insight.cummax(m2, 1).ptr != C_NULL)

# cummin
check("cummin", Insight.cummin(m2, 1).ptr != C_NULL)

# count_nonzero
check("count_nonzero", Insight.count_nonzero(m2).ptr != C_NULL)

# median
check("median", Insight.median(m2).ptr != C_NULL)

# quantile
check("quantile", Insight.quantile(m2, 0.5).ptr != C_NULL)

# nansum
check("nansum", Insight.nansum(m2).ptr != C_NULL)

# nanmean
check("nanmean", Insight.nanmean(m2).ptr != C_NULL)

# nanstd
check("nanstd", Insight.nanstd(m2).ptr != C_NULL)

# sem
check("sem", Insight.sem(m2).ptr != C_NULL)

println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0; exit(1); end
